File: language/typescript/standard-library/data-structures/topologicalSort.ts
/**
 * 手写拓扑排序
 *
 * 考点：
 * - Kahn 算法：反复移除入度为 0 的节点。
 * - 检测环。
 */

export function topologicalSort<T>(graph: Map<T, T[]>): T[] {
  const inDegree = new Map<T, number>();
  for (const [node, neighbors] of graph) {
    if (!inDegree.has(node)) inDegree.set(node, 0);
    for (const neighbor of neighbors) {
      inDegree.set(neighbor, (inDegree.get(neighbor) ?? 0) + 1);
    }
  }

  const queue: T[] = [];
  for (const [node, degree] of inDegree) {
    if (degree === 0) queue.push(node);
  }

  const result: T[] = [];
  while (queue.length > 0) {
    const node = queue.shift()!;
    result.push(node);
    for (const neighbor of graph.get(node) ?? []) {
      const degree = inDegree.get(neighbor)! - 1;
      inDegree.set(neighbor, degree);
      if (degree === 0) queue.push(neighbor);
    }
  }

  if (result.length !== inDegree.size) {
    throw new Error("graph contains a cycle");
  }
  return result;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const tasks = new Map<string, string[]>([
    ["build", ["test"]],
    ["test", ["deploy"]],
    ["deploy", []],
    ["lint", ["test"]],
  ]);
  console.log(topologicalSort(tasks));
}

File: language/python/standard-library/data_structures/topological_sort.py
"""
手写拓扑排序。

考点：
- Kahn 算法： repeatedly remove nodes with zero in-degree.
- 检测环：如果结果长度 != 节点数，则存在环。
"""

from __future__ import annotations

from collections import deque
from collections.abc import Mapping
from typing import TypeVar

T = TypeVar("T")


def topological_sort(graph: Mapping[T, list[T]]) -> list[T]:
    """Return a topological ordering of ``graph``.

    Raises:
        ValueError: If the graph contains a cycle.
    """
    # Compute in-degrees.
    in_degree: dict[T, int] = {node: 0 for node in graph}
    for node in graph:
        for neighbor in graph[node]:
            if neighbor not in in_degree:
                in_degree[neighbor] = 0
            in_degree[neighbor] += 1

    queue: deque[T] = deque(node for node, degree in in_degree.items() if degree == 0)
    result: list[T] = []

    while queue:
        node = queue.popleft()
        result.append(node)
        for neighbor in graph.get(node, []):
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    if len(result) != len(in_degree):
        raise ValueError("graph contains a cycle")

    return result


if __name__ == "__main__":
    tasks = {
        "build": ["test"],
        "test": ["deploy"],
        "deploy": [],
        "lint": ["test"],
    }
    print(topological_sort(tasks))
    # ['build', 'lint', 'test', 'deploy'] (order may vary for independent nodes)

    cyclic = {"A": ["B"], "B": ["A"]}
    try:
        topological_sort(cyclic)
    except ValueError as e:
        print("caught:", e)
